library/core/src/slice/ascii.rs

```diff
@@ -323,100 +315,39 @@ pub const fn is_ascii_simple(mut bytes: &[u8]) -> bool {
-    // have alignment information it should have given a `usize::MAX` for
-    // `align_offset` earlier, sending things through the scalar path instead of
-    // this one, so this check should pass if it's reachable.
-    debug_assert!(word_ptr.is_aligned_to(mem::align_of::<usize>()));
-
-    // Read subsequent words until the last aligned word, excluding the last
-    // aligned word by itself to be done in tail check later, to ensure that
-    // tail is always one `usize` at most to extra branch `byte_pos == len`.
-    while byte_pos < len - USIZE_SIZE {
-        // Sanity check that the read is in bounds
-        debug_assert!(byte_pos + USIZE_SIZE <= len);
-        // And that our assumptions about `byte_pos` hold.
-        debug_assert!(matches!(
-            word_ptr.cast::<u8>().guaranteed_eq(start.wrapping_add(byte_pos)),
-            // These are from the same allocation, so will hopefully always be
-            // known to match even in CTFE, but if it refuses to compare them
-            // that's ok since it's just a debug check anyway.
-            None | Some(true),
-        ));
+const fn is_ascii(bytes: &[u8]) -> bool {
+    // Constant chosen to enable `pmovmskb` instruction on x86-64
+    const N: usize = 32;
+
+    let mut i = 0;
+
+    while i + N <= bytes.len() {
+        let chunk_end = i + N;
+
+        // Get LLVM to produce a `pmovmskb` instruction on x86-64 which
+        // creates a mask from the most significant bit of each byte.
+        // ASCII bytes are less than 128 (0x80), so their most significant
+        // bit is unset. Thus, detecting non-ASCII bytes can be done in one
+        // instruction.
+        let mut count = 0;
+        while i < chunk_end {
+            count += (bytes[i] <= 127) as u8;
+            i += 1;
+        }
 
-        // SAFETY: We know `word_ptr` is properly aligned (because of
-        // `align_offset`), and we know that we have enough bytes between `word_ptr` and the end
-        let word = unsafe { word_ptr.read() };
-        if contains_nonascii(word) {
+        // All bytes should be <= 127 so count is equal to chunk size.
+        if count != N as u8 {
             return false;
         }
-
-        byte_pos += USIZE_SIZE;
-        // SAFETY: We know that `byte_pos <= len - USIZE_SIZE`, which means that
-        // after this `add`, `word_ptr` will be at most one-past-the-end.
-        word_ptr = unsafe { word_ptr.add(1) };
     }
 
-    // Sanity check to ensure there really is only one `usize` left. This should
-    // be guaranteed by our loop condition.
-    debug_assert!(byte_pos <= len && len - byte_pos <= USIZE_SIZE);
-
-    // SAFETY: This relies on `len >= USIZE_SIZE`, which we check at the start.
-    let last_word = unsafe { (start.add(len - USIZE_SIZE) as *const usize).read_unaligned() };
+    // Process the remaining `bytes.len() % N` bytes.
+    let mut is_ascii = true;
+    while i < bytes.len() {
+        is_ascii &= bytes[i] <= 127;
```

Comment:
On the godbolt link it looks like the compiler tries to unroll and autovectorize this loop again, which might not be beneficial for usually low loop counts. The code size would be much smaller if the remainder loop has a simple early exit, but I haven't run the benchmarks on that:

```rust
    // Process the remaining `bytes.len() % N` bytes.
    while i < bytes.len() {
        if bytes[i] > 127 {
            return false;
        }
    }

    true
```

If that turns out to be a slowdown, another thing to try could be a second loop with a smaller chunk size:

```rust
    // Constant chosen to enable `pmovmskb` instruction on x86-64
    const N1: usize = 32;
    const N2: usize = 8; // or 16

    let mut i = 0;

    while i + N1 <= bytes.len() {
        let chunk_end = i + N1;

        let mut count = 0;
        while i < chunk_end {
            count += (bytes[i] <= 127) as u8;
            i += 1;
        }

        // All bytes should be <= 127 so count is equal to chunk size.
        if count != N1 as u8 {
            return false;
        }
    }

    while i + N2 <= bytes.len() {
        let chunk_end = i + N2;

        let mut count = 0;
        while i < chunk_end {
            count += (bytes[i] <= 127) as u8;
            i += 1;
        }

        // All bytes should be <= 127 so count is equal to chunk size.
        if count != N2 as u8 {
            return false;
        }
    }

    // Process the remaining `bytes.len() % N` bytes.
    while i < bytes.len() {
        if bytes[i] > 127 {
            return false;
        }
    }

    true
```

Reply:
> On the godbolt link it looks like the compiler tries to unroll and autovectorize this loop again, which might not be beneficial for usually low loop counts.

I reported this when I had the time to https://github.com/rust-lang/rust/issues/130795

> The code size would be much smaller if the remainder loop has a simple early exit, but I haven't run the benchmarks on that:

This ended up being a slowdown, also note it's missing the `i += 1;` for anyone wanting to experiment with this.
I made that mistake refactoring a few times trying out different loops.

I tried `slice::from_raw_parts` on the tail to use the `is_ascii_simple` function above and that was slower too. It seemed faster to just eat the extra unrolled code since I couldn't get a scalar remainder loop construction that was as fast.

> If that turns out to be a slowdown, another thing to try could be a second loop with a smaller chunk size:

I tried a few permutations of that, and reductions of the constant like 16, 8, 4, 2, 1. The second loop of 8 seemed to perform the best on the benchmark suite, but only performed markedly better on the `unaligned_both_medium` and `unaligned_tail_medium` cases.

```rust
ascii::is_ascii_slice::long::case00_libcore                          24.39/iter +/- 1.16
ascii::is_ascii_slice::long::case04_while_loop_32                     8.14/iter +/- 0.80
ascii::is_ascii_slice::long::case05_while_loop_32_8                   9.28/iter +/- 1.17
ascii::is_ascii_slice::medium::case00_libcore                         2.93/iter +/- 0.06
ascii::is_ascii_slice::medium::case04_while_loop_32                   1.64/iter +/- 0.30
ascii::is_ascii_slice::medium::case05_while_loop_32_8                 1.86/iter +/- 0.09
ascii::is_ascii_slice::short::case00_libcore                          5.34/iter +/- 0.99
ascii::is_ascii_slice::short::case04_while_loop_32                    3.91/iter +/- 0.13
ascii::is_ascii_slice::short::case05_while_loop_32_8                  3.82/iter +/- 0.11
ascii::is_ascii_slice::unaligned_both_long::case00_libcore           23.13/iter +/- 1.18
ascii::is_ascii_slice::unaligned_both_long::case04_while_loop_32      5.29/iter +/- 0.26
ascii::is_ascii_slice::unaligned_both_long::case05_while_loop_32_8    5.31/iter +/- 0.18
ascii::is_ascii_slice::unaligned_both_medium::case00_libcore          3.04/iter +/- 0.35
ascii::is_ascii_slice::unaligned_both_medium::case04_while_loop_32    2.78/iter +/- 0.08
ascii::is_ascii_slice::unaligned_both_medium::case05_while_loop_32_8  1.62/iter +/- 0.10
ascii::is_ascii_slice::unaligned_head_long::case00_libcore           23.03/iter +/- 2.51
ascii::is_ascii_slice::unaligned_head_long::case04_while_loop_32      8.04/iter +/- 0.24
ascii::is_ascii_slice::unaligned_head_long::case05_while_loop_32_8   13.04/iter +/- 0.34
ascii::is_ascii_slice::unaligned_head_medium::case00_libcore          3.78/iter +/- 0.52
ascii::is_ascii_slice::unaligned_head_medium::case04_while_loop_32    6.86/iter +/- 0.61
ascii::is_ascii_slice::unaligned_head_medium::case05_while_loop_32_8  8.39/iter +/- 0.16
ascii::is_ascii_slice::unaligned_tail_long::case00_libcore           24.26/iter +/- 0.22
ascii::is_ascii_slice::unaligned_tail_long::case04_while_loop_32      4.54/iter +/- 0.15
ascii::is_ascii_slice::unaligned_tail_long::case05_while_loop_32_8    4.69/iter +/- 0.05
ascii::is_ascii_slice::unaligned_tail_medium::case00_libcore          2.89/iter +/- 0.40
ascii::is_ascii_slice::unaligned_tail_medium::case04_while_loop_32    2.63/iter +/- 0.09
ascii::is_ascii_slice::unaligned_tail_medium::case05_while_loop_32_8  1.98/iter +/- 0.13
```